### PR TITLE
Refactoring of acc:u0

### PR DESCRIPTION
**Pull request by AcK77**
- Move all account things to the account service
- More accurate IAccountServiceForApplication
- Add helper to UInt128

### CODE REVIEW

**Comment by AcK77** [comment]
Should fix #668 too.

File: Ryujinx.HLE/Utilities/UInt128.cs [hunk 1]
```diff
@@ -26,6 +32,16 @@ public UInt128(string hex)
...
 
+        public bool IsNull()
+        {
+            if (Low == 0 && High == 0)
+            {
+                return true;
+            }
+
+            return false;
+        }
```
Comment (FearlessTobi) [inline Ryujinx.HLE/Utilities/UInt128.cs]: Can't this just be "return Low == 0 && High == 0"?

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 2]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
-                Logger.PrintWarning(LogClass.ServiceAcc, $"User 0x{uuid} not found!");
+                Logger.PrintWarning(LogClass.ServiceAcc, $"User 0x{userId} not found!");
 
                 return MakeError(ErrorModule.Account, AccErr.UserNotFound);
             }
 
-            MakeObject(context, new IProfile(profile));
+            MakeObject(context, new IProfile(userProfile));
+
+            // Don't occur in our case.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: This doesn't seem like correct English to me

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 3]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
 
         // IsUserRegistrationRequestPermitted(u64, pid) -> bool
         public long IsUserRegistrationRequestPermitted(ServiceCtx context)
         {
-            long unknown = context.RequestData.ReadInt64();
-
-            Logger.PrintStub(LogClass.ServiceAcc, new { unknown });
-
-            context.ResponseData.Write(false);
+            // u64 don't seems to be used in RE.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 4]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
 
-            Logger.PrintStub(LogClass.ServiceAcc, new { unknown });
+                return 0;
+            }
+
+            bool baasCheck = context.RequestData.ReadBoolean();
+
+            if (baasCheck)
+            {
+                // It check something related to baas (online) and then return an invalid UserId if the check in baas return an error code.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito.

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 5]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
 
-            profile.Uuid.Write(context.ResponseData);
+            // Since we return an invalid UserId if there is more than one user, we can return the first one.
+            context.Device.System.State.Account.GetFirst().UserId.Write(context.ResponseData);
 
             return 0;
         }
 
         // InitializeApplicationInfo(u64, pid)
+        // Both call use the same submethod in 6.0.0+, maybe changes are further, in the arp:r call ?
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 6]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
         public long InitializeApplicationInfo(ServiceCtx context)
         {
+            if (_applicationLaunchProperty != null)
+            {
+                return MakeError(ErrorModule.Account, AccErr.ApplicationLaunchPropertyAlreadyInit);
+            }
+
+            // u64 don't seems to be used in RE.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 7]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
             long unknown = context.RequestData.ReadInt64();
 
+            // TODO: It call nn::arp::detail::IReader::GetApplicationLaunchProperty() with the current pid and return an ApplicationLaunchProperty.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 8]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
+            }
+
+            if (_applicationLaunchProperty == null)
+            {
+                return MakeError(ErrorModule.Account, AccErr.InvalidArgument);
+            }
+
+            MakeObject(context, new IManagerForApplication(userId, _applicationLaunchProperty));
+
+            // Don't occur in our case.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 9]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
+        {
+            // TODO : Validate the following check.
+            /*
+            if (_applicationLaunchProperty != null)
+            {
+                return MakeError(ErrorModule.Account, AccErr.ApplicationLaunchPropertyAlreadyInit);
+            }
+            */
+
+            // It call nn::arp::detail::IReader::GetApplicationControlProperty() to get the NACP file.
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Dito

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 10]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
+                    BaseGameStorageId   = 0x03;
+                    UpdateGameStorageId = 0x00;
+                }
+
+                return MakeError(ErrorModule.Account, AccErr.InvalidArgument);
+            }
+            else
+            */
+            {
+                _applicationLaunchProperty = new ApplicationLaunchProperty
```
Comment (FearlessTobi) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Maybe I'm just stupid, but where is this actually used?

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 11]
```diff
@@ -18,67 +25,90 @@ public IAccountService()
...
+              //{ 141, ListQualifiedUsers                   }, // 6.0.0+
+                { 150, IsUserAccountSwitchLocked            }, // 6.0.0+
             };
         }
 
         // GetUserCount() -> i32
         public long GetUserCount(ServiceCtx context)
         {
-            context.ResponseData.Write(context.Device.System.State.GetUserCount());
+            context.ResponseData.Write(context.Device.System.State.Account.GetUserCount());
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Would be nice if the `AccountState` was part of the account service. `context.Device.System.State.Account` is already getting a bit too far imo.

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 2 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Maybe `Doesn't occur in our case`.

Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: Btw if it never happens, then I don't think we need the commented out error return.

File: Ryujinx.HLE/Utilities/UInt128.cs (hunk 1 above)
Comment (BaronKiko) [reply Ryujinx.HLE/Utilities/UInt128.cs]: Rather `public bool IsNull() => ((Low == 0) && (High == 0))`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 3 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `The u64 argument seems to be unused.`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 4 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `It checks for something related to baas (online) and then returns an invalid UserId if an error code is returned from the check in baas.`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 12]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
-            Logger.PrintStub(LogClass.ServiceAcc, new { unknown });
+                return 0;
+            }
+
+            bool baasCheck = context.RequestData.ReadBoolean();
+
+            if (baasCheck)
+            {
+                // It check something related to baas (online) and then return an invalid UserId if the check in baas return an error code.
+                // In our case, we can just log it for now.
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `This is not properly implemented yet, so just log it for now.`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 13]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
 
-            UserProfile profile = context.Device.System.State.LastOpenUser;
+                Logger.PrintStub(LogClass.ServiceAcc, new { baasCheck });
+            }
 
-            profile.Uuid.Write(context.ResponseData);
+            // Since we return an invalid UserId if there is more than one user, we can return the first one.
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: This comment is confusing imo. Should it return a invalid user id, or the user id of the first user?

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 5 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `Both calls uses the same functions on 6.0.0+, maybe there's something different further along, in the arp:r call?`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 6 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `The u64 argument seems to be unused.`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 7 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `call` -> `calls`, `return` -> `returns`. Also, maybe capitalize `pid`.

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 8 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `Doesn't`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 9 above)
Comment (gdkchan) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `calls`

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs [hunk 14]
```diff
@@ -87,76 +117,212 @@ private long WriteUserList(ServiceCtx context, IEnumerable<UserProfile> profiles
...
+            // TODO : Validate the following check.
+            /*
+            if (_applicationLaunchProperty != null)
+            {
+                return MakeError(ErrorModule.Account, AccErr.ApplicationLaunchPropertyAlreadyInit);
+            }
+            */
+
+            // It call nn::arp::detail::IReader::GetApplicationControlProperty() to get the NACP file.
+            // But since we use LibHac and we load one Application at a time. It's not needed.
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: `at a time. It's not needed.` -> `at a time, it's not necessary.`

File: Ryujinx.HLE/Utilities/UInt128.cs (hunk 1 above)
Comment (gdkchan) [reply Ryujinx.HLE/Utilities/UInt128.cs]: `public bool IsNull() => (Low == 0) && (High == 0);` is better imo. Maybe its better to use a property for that.
```cs
public bool IsNull => (Low == 0) && (High == 0);
```

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 10 above)
Comment (AcK77) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: in `GetBaasAccountManagerForApplication()` just above

File: Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs (hunk 11 above)
Comment (AcK77) [reply Ryujinx.HLE/HOS/Services/Acc/IAccountService.cs]: That's should be done when the profile manager will be coded

**Pull request merged**